### PR TITLE
Refactor/of the ai code and architecture zappy ai

### PR DESCRIPTION
**Pull request by theophile-jr**
Fixed the merge with child-ai branch, put comments on all the missing functions, correctly merged marius & child ai functions. The refactor is lighter than we wanted.

### CODE REVIEW

File: ai/src/AI.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 ## AI
 ##
 
+
 import os
```
Comment (mathieurobert1) [inline ai/src/AI.py:9]:
```suggestion
##

import os
```

File: ai/src/Player/Player.py [hunk 2]
```diff
@@ -585,78 +604,24 @@ def lookingForFood(self):
...
-                break
+        (found, index) = self.foodInVision(self.vision)
+        if not found:
+            return random.choice([self.moveForward, self.moveForward, self.turnRight, self.turnLeft])()
+        self.goGetItem(index, [Item.FOOD] * self.vision[index].food)
         self.cmdInventory()
 
-
-    def lookingForStones(self):
+    def lookingForStones(self): #UP TO FIX YKYK
```
Comment (mathieurobert1) [inline ai/src/Player/Player.py:613]:
```suggestion
    def lookingForStones(self): 
```

File: ai/src/Player/Player.py [hunk 3]
```diff
@@ -877,3 +841,82 @@ def chooseAction(self):
...
+            self.turnLeft()
+        for i in range(x):
+            self.moveForward()
+        for i in range(0, len(itemSeek)):
+            if len(self.actions) < 9:
+                self.take(itemSeek[i].value)
+
+    
+
+    def foodInVision(self, vision : list):
```
Comment (mathieurobert1) [inline ai/src/Player/Player.py:874]:
```suggestion
                self.take(itemSeek[i].value)

    def foodInVision(self, vision : list):
```

File: ai/src/Utils/Utils.py [hunk 4]
```diff
@@ -23,3 +23,66 @@ def stringifyData(data : str):
             string += c
     string += "\""
     return string
+
+
+def mapRangeOpti(n):
```
Comment (mathieurobert1) [inline ai/src/Utils/Utils.py:28]:
```suggestion
    return string

def mapRangeOpti(n):
```

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
Some requests but overall it's fine :+1:

**Review by mathieurobert1 — DISMISSED** [review]
:+1:

File: ai/src/Player/Player.py [hunk 5]
```diff
@@ -5,13 +5,30 @@
...
+    FOOD = 0
+    STONES = 1
+    FORKING = 2
+    BROADCASTING = 3
+    HANDLINGRESPONSE = 4
+    WAITING = 5
+    ELEVATING = 6
+    REGROUP = 7
+    DROPPING = 8
+    NONE = 9
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:31]: There is a file with this code in src/enum/

File: ai/src/Player/Player.py [hunk 6]
```diff
@@ -5,13 +5,30 @@
 ## Player
 ##
 
+from enum import Enum
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:8]: Can be deleted since the enum is on another file

File: ai/src/AI.py [hunk 7]
```diff
@@ -102,7 +102,6 @@ def serverCommunicationInThread(self):
                 self.player.commands.pop(0)
                 self.player.callbacks.pop(0)
 
-
```
Comment (Marius-P1) [inline ai/src/AI.py:105]: Don't remove the space, cause between all the function there is 2 new lines (or update for all the functions but the code will be more difficult to read)

File: ai/src/Player/Player.py [hunk 8]
```diff
@@ -143,9 +160,14 @@ class Player:
...
             Choose the action of the player
+        goGetItem()
+            Go get items at tile
+        foodInVision()
+            Check if there is food in the vision
+        stonesInVision()
+            Check if there are stones in the vision
     """
 
-
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:148]: Don't remove the space, cause between all the function there is 2 new lines (or update for all the functions but the code will be more difficult to read)

File: ai/src/Player/Player.py [hunk 9]
```diff
@@ -423,7 +446,6 @@ def updateEjectionReceived(self, message : str):
         direction = int(message)
         self.ejectionReceived.append(direction)
 
-
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:426]: Don't remove the space, cause between all the function there is 2 new lines (or update for all the functions but the code will be more difficult to read)

File: ai/src/Player/Player.py [hunk 10]
```diff
@@ -512,7 +534,6 @@ def handleResponse(self, response : str):
             if response == "ok":
                 self.broadcastReceived = []
 
-
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:515]: Don't remove the space, cause between all the function there is 2 new lines (or update for all the functions but the code will be more difficult to read)

File: ai/src/Player/Player.py [hunk 11]
```diff
@@ -539,7 +560,6 @@ def updateModeSlave(self):
         elif self.inventory.food >= 45:
             self.currentMode = Mode.STONES
 
-
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:542]: Don't remove the space, cause between all the function there is 2 new lines (or update for all the functions but the code will be more difficult to read)

File: ai/src/Player/Player.py [hunk 12]
```diff
@@ -565,7 +585,6 @@ def updateModeLeader(self):
                 self.currentMode = Mode.WAITING
         self.currentFood = self.inventory.food
 
-
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:568]: Don't remove the space, cause between all the function there is 2 new lines (or update for all the functions but the code will be more difficult to read)

File: ai/src/Player/Player.py [hunk 13]
```diff
@@ -666,7 +631,6 @@ def askSlavesForInventory(self):
         self.broadcast("Inventory")
         self.nbSlaves = 0
 
-
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:669]: Don't remove the space, cause between all the function there is 2 new lines (or update for all the functions but the code will be more difficult to read)

File: ai/src/Player/Player.py [hunk 14]
```diff
@@ -585,78 +604,24 @@ def lookingForFood(self):
...
-                self.take("food")
-            else:
-                break
+        (found, index) = self.foodInVision(self.vision)
+        if not found:
+            return random.choice([self.moveForward, self.moveForward, self.turnRight, self.turnLeft])()
+        self.goGetItem(index, [Item.FOOD] * self.vision[index].food)
         self.cmdInventory()
 
-
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:616]: Don't remove the space, cause between all the function there is 2 new lines (or update for all the functions but the code will be more difficult to read)

**Review by Marius-P1 — CHANGES_REQUESTED** [review]
Some changes :+1:

**Review by theophile-jr — COMMENTED** [review]
Changes following marius's requests changes.

File: ai/src/Player/Player.py [hunk 15]
```diff
@@ -5,13 +5,17 @@
...
 ##
 
+import random
 import sys
+from typing import List
 
 from ai.src.Enum.Mode import Mode
 from ai.src.Enum.Role import Role
 from ai.src.Enum.Action import Action
+from ai.src.Enum.Item import Item
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:15]:
```suggestion
from ai.src.Enum.Item import Item
from ai.src.Enum.Action import Action
```

File: ai/src/Player/Player.py [hunk 16]
```diff
@@ -877,3 +836,83 @@ def chooseAction(self):
...
+                List of items in view
+        
+        Returns : 
+            truple: bool for found stones, tile's index, list of stones enum
+        """
+        foundStones : List[Item] = []
+
+        map = list(enumerate(vision))
+        map[0], map[1] = map[1], map[0]
+        for i, v in map:
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:903]: Why don't you use the same method as line 882 (using mapRangeOpti) ?

File: ai/src/Player/Player.py [hunk 17]
```diff
@@ -877,3 +836,83 @@ def chooseAction(self):
...
+                foundStones.append(Item.SIBUR)
+            if v.mendiane > 0:
+                foundStones.append(Item.MENDIANE)
+            if v.phiras > 0:
+                foundStones.append(Item.PHIRAS)
+            if v.thystame > 0:
+                foundStones.append(Item.THYSTAME)
+            if (len(foundStones) > 0):
+                return (True, i, foundStones)
+        return (False, -1, None)
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:918]: Here you take stones from the nearest case where stones are present but not from the case where there is the more diversity of stone like i was doing in looking for stones. So maybe you can update that so the AI will be able to catch different stones (more diversity = more chance to take the stone we needed)

File: ai/src/Player/Player.py
Comment (Marius-P1) [inline ai/src/Player/Player.py:1]: Can you put your method on top or behind the method they are use (like lookingForFood) so it won't break the reasoning order i have done for the choose action ?

File: ai/src/Utils/Utils.py [hunk 18]
```diff
@@ -23,3 +23,66 @@ def stringifyData(data : str):
...
+        width : int
+            width of the current row
+
+    Returns :
+        int : the number of movements to do
+    """
+    if middle == target:
+        return 0
+    return target - middle
+
```
Comment (Marius-P1) [inline ai/src/Utils/Utils.py:60]:
```suggestion


```

File: ai/src/Utils/Utils.py [hunk 19]
```diff
@@ -23,3 +23,66 @@ def stringifyData(data : str):
...
+            n : int
+                Vision length
+            Returns :
+                List of indexes
+    """
+    yield 1
+    yield 0
+    for i in range(2, n):
+        yield i
+
```
Comment (Marius-P1) [inline ai/src/Utils/Utils.py:41]:
```suggestion


```

**Review by Marius-P1 — CHANGES_REQUESTED** [review]
Some little changes but overall is great :+1:

File: ai/src/Player/Player.py (hunk 16 above)
Comment (theophile-jr) [reply ai/src/Player/Player.py:903]: because this is an enumerate (rendering a list of tuple) and not a range ;)

**Review by mathieurobert1 — APPROVED** [review]

**Review by mathieurobert1 — APPROVED** [review]

File: .gitkeep
Comment (Marius-P1) [inline .gitkeep:1]: Why ?

File: ai/src/Player/Player.py [hunk 20]
```diff
@@ -615,84 +626,118 @@ def updateMode(self):
...
-            self.cmdInventory()
+        total : int = 0
+
+        for i in mapRangeOpti(len(vision)):
+            if vision[i].food > 0:
+                return (True, i)
+        return (False, -1)
+
+
+    def goGetItem(self, index, itemSeek : List[Item]):
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:648]: if you write the type of one param, write the type of the other too

**Review by Marius-P1 — CHANGES_REQUESTED** [review]
Two little change but the overall is great

File: .gitkeep
Comment (theophile-jr) [reply .gitkeep:1]: mistake, removed

**Review by mathieurobert1 — APPROVED** [review]
If @Marius-P1 accept, it's ok.

**Review by Marius-P1 — DISMISSED** [review]

**Review by mathieurobert1 — APPROVED** [review]

File: ai/src/Player/Player.py [hunk 21]
```diff
@@ -5,15 +5,19 @@
 ## Player
 ##
 
+import random
```
Comment (Marius-P1) [inline ai/src/Player/Player.py:8]: double import
```suggestion
```

**Review by Marius-P1 — APPROVED** [review]
:ok:

**Review by mathieurobert1 — APPROVED** [review]

**Pull request merged**